src/HotChocolate/ApolloFederation/src/ApolloFederation/EntityTypeInterceptor.cs

```diff
@@ -0,0 +1,117 @@
+using HotChocolate.Types;
+using HotChocolate.Types.Descriptors;
+using HotChocolate.Types.Descriptors.Definitions;
+using static HotChocolate.ApolloFederation.WellKnownContextData;
+
+namespace HotChocolate.ApolloFederation
+{
+    internal class EntityTypeInterceptor : TypeInterceptor
+    {
+        private readonly List<ObjectType> _entityTypes = new List<ObjectType>();
+
+        public override bool TriggerAggregations { get; } = true;
+
+        public override void OnAfterInitialize(
+            ITypeDiscoveryContext discoveryContext,
+            DefinitionBase definition,
+            IDictionary<string, object> contextData)
+        {
+            AddToUnionIfHasTypeLevelKeyDirective(
+                discoveryContext,
+                definition);
+
+            AggregatePropertyLevelKeyDirectives(
+                discoveryContext,
+                definition);
+        }
+
+        public override void OnBeforeCompleteType(
+            ITypeCompletionContext completionContext,
+            DefinitionBase definition,
+            IDictionary<string, object> contextData) =>
+            AddMemberTypesToTheEntityUnionType(
+                completionContext,
+                definition);
+
+        private void AddToUnionIfHasTypeLevelKeyDirective(
+            ITypeDiscoveryContext discoveryContext,
+            DefinitionBase definition)
+        {
+            if (discoveryContext.Type is ObjectType objectType &&
+                definition is ObjectTypeDefinition objectTypeDefinition)
+            {
+                if (objectTypeDefinition.Directives.Any(
+                        d => d.Reference is NameDirectiveReference { Name: { Value: "key" }}) ||
+                    objectTypeDefinition.Fields.Any(
+                        f => f.ContextData.ContainsKey("key")))
+                {
+                    _entityTypes.Add(objectType);
+                }
+            }
+        }
+
+        private void AggregatePropertyLevelKeyDirectives(
+            ITypeDiscoveryContext discoveryContext,
+            DefinitionBase definition)
+        {
+            if (discoveryContext.Type is ObjectType objectType &&
+                definition is ObjectTypeDefinition objectTypeDefinition)
+            {
+                // if we find key markers on our fields, we need to construct the key directive
+                // from the annotated fields.
+                if (objectTypeDefinition.Fields.Any(f => f.ContextData.ContainsKey(KeyMarker)))
+                {
+                    IReadOnlyList<ObjectFieldDefinition> fields = objectTypeDefinition.Fields;
+                    var fieldSet = new StringBuilder();
+
+                    for (var i = 0; i < fields.Count; i++)
+                    {
+                        ObjectFieldDefinition fieldDefinition = fields[i];
+                        if (fieldDefinition.ContextData.ContainsKey(KeyMarker))
+                        {
+                            if (i > 0)
```

Comment:
In case the first field does not contain a field marker then the ' ' is still appended. I think this should be fieldset.Length > 0

Reply:
How do you mean that? What we want is this

`field_a field_b field_c`

so if i > 0 means that field_a in this case is already added, meaning I have to append a space.

Reply:
you are right!